File: swift/Sources/SignalClient/zkgroup/ProfileKey.swift
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class ProfileKey: ByteArray {

  public static let SIZE: Int = 32

  public required init(contents: [UInt8]) throws {
    try super.init(newContents: contents, expectedLength: ProfileKey.SIZE)
  }

  public func getCommitment(uuid: UUID) throws -> ProfileKeyCommitment {
    return try withUnsafePointerToSerialized { contents in
      try withUnsafePointer(to: uuid.uuid) { uuid in
        try invokeFnReturningSerialized {
          signal_profile_key_get_commitment($0, contents, uuid)
        }
      }
    }
  }

  public func getProfileKeyVersion(uuid: UUID) throws -> ProfileKeyVersion {
    return try withUnsafePointerToSerialized { contents in
      try withUnsafePointer(to: uuid.uuid) { uuid in
        try invokeFnReturningSerialized {
          signal_profile_key_get_profile_key_version($0, contents, uuid)
        }
      }
    }
  }

}

File: java/shared/java/org/signal/libsignal/zkgroup/profiles/ProfileKey.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.profiles;

import org.signal.libsignal.protocol.ServiceId.Aci;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.VerificationFailedException;
import org.signal.libsignal.zkgroup.internal.ByteArray;
import org.signal.libsignal.internal.Native;

public final class ProfileKey extends ByteArray {

  public ProfileKey(byte[] contents) throws InvalidInputException {
    super(contents);
    Native.ProfileKey_CheckValidContents(contents);
  }

  public ProfileKeyCommitment getCommitment(Aci userId) {
    byte[] newContents = Native.ProfileKey_GetCommitment(contents, userId.toServiceIdFixedWidthBinary());

    try {
      return new ProfileKeyCommitment(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ProfileKeyVersion getProfileKeyVersion(Aci userId) {
    byte[] newContents = Native.ProfileKey_GetProfileKeyVersion(contents, userId.toServiceIdFixedWidthBinary());

    try {
      return new ProfileKeyVersion(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public byte[] deriveAccessKey() {
    return Native.ProfileKey_DeriveAccessKey(contents);
  }

}
